src/aws-elasticache/user-group.ts

```diff
@@ -0,0 +1,167 @@
+import { IResource, Lazy, Names, Resource, Stack, Token, aws_elasticache } from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+import { IUser } from './user';
+import { Engine } from './util';
+
+/**
+ * An User Group
+ */
+export interface IUserGroup extends IResource {
+  /**
+   * The ARN of the user group.
+   *
+   * @attribute
+   */
+  readonly userGroupArn: string;
+  /**
+   * The ID of the user group.
+   *
+   * @attribute
+   */
+  readonly userGroupId: string;
+}
+
+/**
+ * Properties for defining an User Group.
+ */
+export interface UserGroupProps {
+  /**
+   * The ID of the user group.
+   *
+   * \`userGroupId\` can have up to 40 characters.
+   *
+   * \`userGroupId\` must consist only of alphanumeric characters or hyphens,
+   * with the first character as a letter, and it can't end with a hyphen or contain two consecutive hyphens.
+   *
+   * @default - auto generate
+   */
+  readonly userGroupId?: string;
+
+  /**
+   * The list of User that belong to the user group.
+   *
+   * `default` user is automatically added to the user group.
+   *
+   * @default - only `default` user added
+   */
+  readonly users?: IUser[];
+}
+
+/**
+ * Attributes for importing an User Group.
+ */
+export interface UserGroupAttributes {
+  /**
+   * The ID of the user group.
+   */
+  readonly userGroupId: string;
+}
+
+/**
+ * Represents an user group construct in AWS CDK.
+ *
+ * @example
+ * declare const user: User;
+ *
+ * const userGroup = new UserGroup(
+ *   stack,
+ *   'UserGroup',
+ *   {
+ *      user: [user],
+ *   },
+ * );
+ */
+export class UserGroup extends Resource implements IUserGroup {
+  /**
+   * Imports an existing user group from attributes
+   */
+  public static fromUserGroupId(scope: Construct, id: string, userGroupId: string): IUserGroup {
+    class Import extends Resource implements IUserGroup {
+      public readonly userGroupId = userGroupId;
+      public readonly userGroupArn = Stack.of(this).formatArn({
+        service: 'elasticache',
+        resource: 'usergroup',
+        resourceName: userGroupId,
+      });
+    }
+    return new Import(scope, id);
+  }
+
+  /**
+   * The ARN of the user group.
+   */
+  readonly userGroupArn: string;
+
+  /**
+   * The ID of the user group.
+   */
+  readonly userGroupId: string;
+
+  private readonly props: UserGroupProps;
+
+  private readonly users: IUser[];
+
+  constructor(scope: Construct, id: string, props: UserGroupProps) {
+    super(scope, id, {
+      physicalName:
+        props.userGroupId ??
+        Lazy.string({
+          produce: () => Names.uniqueResourceName(this, { maxLength: 40, allowedSpecialCharacters: '-' }).toLowerCase(),
```

Comment:
I prefer to add `separator :'-'` because the generated name will be more readable in lowercased letters. For example:
without separator: `stacknameconstructpathsrandomhash`
with separator:  `stackname-constructpaths-randomhash`

I know it is a matter of preference, but what do you think?

Reply:
Thank you. I like your suggestion, so I will incorporate.

Reply:
updated.